rfcs/002-merge-types-terms-syntax.md

```diff
@@ -0,0 +1,452 @@
+  to take it into account. But this implies giving semantics to things like
+  `record.is_empty { ; a}` or `{ ; a} & { ; b}`. Those are actually
+  interesting questions outside of the context of this RFC (see
+  [201](https://github.com/tweag/nickel/issues/201)).
+
+  While allowing tails in arbitrary record may look more uniform, it requires
+  more design thinking. There isn't any obvious practical usage that seems to
+  require it today. In practice we expect a record with a tail to actually be a
+  record type and be used as a type annotation or applied as a contract derived
+  from a type, but nothing else. Finally, we can always allow it later on, as it
+  should (or, rather, must) be backward-compatible with the simple solution.
+
+  In consequence, we choose to only allow tail on a record that is actually a
+  type in the current syntax. In fact we can allow for slightly more room here
+  later, so we abstract this notion in the predicate `is_record_type : UniTerm
+  -> Bool` and in the end, **propose to set the syntax to require `r = {decl1,
+  ..., decln; tail} => is_record_type({decl1, ..., decln})`**. In this RFC,
+  we'll define `is_record_type(t)` in the obvious way such as if `t` is a record
+  type in the current syntax, i.e. `is_record_type(t) <=> exists. T : Type, T
+  record type, t = !T`. In particular `{foo = fun x => x; a}` is now rejected.
+
+The `{_: Num}` syntax is not ambiguous and can be special cased, as it is
+already in the current `Type` syntax (it only requires a one character
+look-ahead to distinguish).
+
+#### Enums
+
+With the first release in mind, we propose to just disable support for enum
+altogether. This is a handy but hardly fundamental feature, and this lets us
+more time to find a good syntax replacement. We can also disable `switch`
+temporarily, as currently its only usage is for enums.
+
+### Translation
+
+Let's now define the `term` and `type` translations. Our design principle is to
+try to have them just be simple syntactic translations that insert the implicit
+`#` and `§` appropriately, without having to recurse into terms or do anything
+complex, whenever possible.
+
+The two functions must satisfy the following coherence laws with `! : Term +
+Type -> UniTerm` the embedding of the current syntax to the new unified syntax
+(erasing `#` and `§`, and changing `|` for tails to `;`) and `~` some flavour of
+operational equivalence (we won't prove those rules formally, as this a mere
+design guideline, so we also restrain from giving a formal definition for `~`).
+
+1. For all `t: Term`, `term(!t) ~ t` and for all `T: Type`, `type(!T) = T` .
+   That is, a term or a type in the current syntax with `#` removed will be
+   interpreted the same in the new syntax.
+2. For all `T: Type`, `term(!T) ~ §T`. That is, the interpretation of something
+   that is a type in the current syntax as a term in the new syntax must behave
+   the same as the corresponding contract.
```

Comment:
Isn't there an equivalent law for `#`?

Reply:
I think it is too strict to ask, and is e.g. violated by the fact that currently `{foo : Num}` is a valid term. Thus, according the a dual hypothetical law for `#`, we should have `type({foo : Num}) = #{foo : Num}`. But we certainly want `type({foo: Num}) = {foo : Num}`. Maybe less artificially, and although this is mentioned as a future improvement only, we may want to extract static type information for other terms too, such as `type({foo | Num}) = {foo : Num}`, which would also violate this law.

Reply:
Thinking about this again, we may ask for a weaker law, that is `term(!type(!t)) ~ t` for `t : Term`. Unfortunately, the quirk around `{foo | Num}` and `{foo : Num}` behaving slightly differently is probably breaking this law. We may however mention it as a goal, as the issue is in the current state of the implementation, and is not inherent to this RFC.